File: crates/hydra-cli/src/tui/project.rs
/// Project detection and file awareness.
///
/// Auto-detects project type by scanning the working directory for
/// marker files (Cargo.toml, package.json, go.mod, pyproject.toml, etc.)
/// and provides project metadata for the TUI sidebar and developer commands.

use std::path::{Path, PathBuf};

/// Detected project kind.
#[derive(Clone, Debug, PartialEq)]
pub enum ProjectKind {
    Rust,
    Node,
    Python,
    Go,
    Unknown,
}

impl ProjectKind {
    pub fn label(&self) -> &'static str {
        match self {
            Self::Rust => "Rust",
            Self::Node => "Node.js",
            Self::Python => "Python",
            Self::Go => "Go",
            Self::Unknown => "Unknown",
        }
    }

    pub fn icon(&self) -> &'static str {
        match self {
            Self::Rust => "🦀",
            Self::Node => "⬢",
            Self::Python => "🐍",
            Self::Go => "⚙",
            Self::Unknown => "📁",
        }
    }

    /// The test command for this project type.
    pub fn test_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["test"]),
            Self::Node => ("npm", &["test"]),
            Self::Python => ("python", &["-m", "pytest"]),
            Self::Go => ("go", &["test", "./..."]),
            Self::Unknown => ("echo", &["No test command configured"]),
        }
    }

    /// The build command for this project type.
    pub fn build_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["build"]),
            Self::Node => ("npm", &["run", "build"]),
            Self::Python => ("python", &["-m", "build"]),
            Self::Go => ("go", &["build", "./..."]),
            Self::Unknown => ("echo", &["No build command configured"]),
        }
    }

    /// The run command for this project type.
    pub fn run_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["run"]),
            Self::Node => ("npm", &["start"]),
            Self::Python => ("python", &["main.py"]),
            Self::Go => ("go", &["run", "."]),
            Self::Unknown => ("echo", &["No run command configured"]),
        }
    }

    /// The lint command for this project type.
    pub fn lint_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["clippy"]),
            Self::Node => ("npx", &["eslint", "."]),
            Self::Python => ("python", &["-m", "ruff", "check", "."]),
            Self::Go => ("golangci-lint", &["run"]),
            Self::Unknown => ("echo", &["No lint command configured"]),
        }
    }

    /// The format command for this project type.
    pub fn fmt_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["fmt"]),
            Self::Node => ("npx", &["prettier", "--write", "."]),
            Self::Python => ("python", &["-m", "ruff", "format", "."]),
            Self::Go => ("gofmt", &["-w", "."]),
            Self::Unknown => ("echo", &["No format command configured"]),
        }
    }

    /// The bench command for this project type.
    pub fn bench_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["bench"]),
            Self::Node => ("npm", &["run", "bench"]),
            Self::Python => ("python", &["-m", "pytest", "--benchmark-only"]),
            Self::Go => ("go", &["test", "-bench=.", "./..."]),
            Self::Unknown => ("echo", &["No bench command configured"]),
        }
    }

    /// The doc command for this project type.
    pub fn doc_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["doc", "--open"]),
            Self::Node => ("npx", &["typedoc"]),
            Self::Python => ("python", &["-m", "pdoc", "--html", "."]),
            Self::Go => ("godoc", &["-http=:6060"]),
            Self::Unknown => ("echo", &["No doc command configured"]),
        }
    }

    /// The deps command for this project type.
    pub fn deps_cmd(&self) -> (&'static str, &'static [&'static str]) {
        match self {
            Self::Rust => ("cargo", &["tree", "--depth", "1"]),
            Self::Node => ("npm", &["ls", "--depth=0"]),
            Self::Python => ("pip", &["list"]),
            Self::Go => ("go", &["list", "-m", "all"]),
            Self::Unknown => ("echo", &["No deps command configured"]),
        }
    }
}

/// Detected project information.
#[derive(Clone, Debug)]
pub struct ProjectInfo {
    pub kind: ProjectKind,
    pub root: PathBuf,
    pub name: String,
    pub crate_count: Option<usize>,
    pub git_branch: Option<String>,
    pub git_ahead: Option<usize>,
    pub git_behind: Option<usize>,
}

/// Detect the project type from the given directory.
pub fn detect_project(dir: &Path) -> Option<ProjectInfo> {
    if !dir.is_dir() {
        return None;
    }

    let kind = if dir.join("Cargo.toml").exists() {
        ProjectKind::Rust
    } else if dir.join("package.json").exists() {
        ProjectKind::Node
    } else if dir.join("go.mod").exists() {
        ProjectKind::Go
    } else if dir.join("pyproject.toml").exists()
        || dir.join("setup.py").exists()
        || dir.join("requirements.txt").exists()
    {
        ProjectKind::Python
    } else {
        return None;
    };

    let name = detect_project_name(dir, &kind);
    let crate_count = if kind == ProjectKind::Rust {
        count_rust_crates(dir)
    } else {
        None
    };
    let (git_branch, git_ahead, git_behind) = detect_git_info(dir);

    Some(ProjectInfo {
        kind,
        root: dir.to_path_buf(),
        name,
        crate_count,
        git_branch,
        git_ahead,
        git_behind,
    })
}

fn detect_project_name(dir: &Path, kind: &ProjectKind) -> String {
    match kind {
        ProjectKind::Rust => {
            if let Ok(content) = std::fs::read_to_string(dir.join("Cargo.toml")) {
                // Try workspace name first, then package name
                for line in content.lines() {
                    let line = line.trim();
                    if line.starts_with("name") {
                        if let Some(val) = line.split('=').nth(1) {
                            let name = val.trim().trim_matches('"');
                            if !name.is_empty() {
                                return name.to_string();
                            }
                        }
                    }
                }
            }
        }
        ProjectKind::Node => {
            if let Ok(content) = std::fs::read_to_string(dir.join("package.json")) {
                if let Ok(json) = serde_json::from_str::<serde_json::Value>(&content) {
                    if let Some(name) = json.get("name").and_then(|v| v.as_str()) {
                        return name.to_string();
                    }
                }
            }
        }
        ProjectKind::Go => {
            if let Ok(content) = std::fs::read_to_string(dir.join("go.mod")) {
                if let Some(line) = content.lines().next() {
                    if let Some(module) = line.strip_prefix("module ") {
                        let name = module.rsplit('/').next().unwrap_or(module);
                        return name.to_string();
                    }
                }
            }
        }
        _ => {}
    }

    // Fall back to directory name
    dir.file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("project")
        .to_string()
}

fn count_rust_crates(dir: &Path) -> Option<usize> {
    let content = std::fs::read_to_string(dir.join("Cargo.toml")).ok()?;
    let count = content
        .lines()
        .filter(|l| {
            let trimmed = l.trim();
            trimmed.starts_with('"') && trimmed.contains("crates/")
        })
        .count();
    if count > 0 {
        Some(count)
    } else {
        None
    }
}

fn detect_git_info(dir: &Path) -> (Option<String>, Option<usize>, Option<usize>) {
    let branch = std::process::Command::new("git")
        .args(["rev-parse", "--abbrev-ref", "HEAD"])
        .current_dir(dir)
        .output()
        .ok()
        .and_then(|o| {
            if o.status.success() {
                Some(String::from_utf8_lossy(&o.stdout).trim().to_string())
            } else {
                None
            }
        });

    let (ahead, behind) = if branch.is_some() {
        let output = std::process::Command::new("git")
            .args(["rev-list", "--left-right", "--count", "HEAD...@{upstream}"])
            .current_dir(dir)
            .output()
            .ok();
        if let Some(o) = output {
            if o.status.success() {
                let s = String::from_utf8_lossy(&o.stdout);
                let parts: Vec<&str> = s.trim().split('\t').collect();
                if parts.len() == 2 {
                    let a = parts[0].parse::<usize>().ok();
                    let b = parts[1].parse::<usize>().ok();
                    (a, b)
                } else {
                    (None, None)
                }
            } else {
                (None, None)
            }
        } else {
            (None, None)
        }
    } else {
        (None, None)
    };

    (branch, ahead, behind)
}

// File listing, reading, language detection, and git commands
// are in the sibling module `project_files`.
pub use super::project_files::*;

File: crates/hydra-tui/internal/app/project.go
package app

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// ProjectKind identifies the type of project.
type ProjectKind int

const (
	ProjectUnknown ProjectKind = iota
	ProjectRust
	ProjectNode
	ProjectGo
	ProjectPython
)

// ProjectInfo holds detected project information.
type ProjectInfo struct {
	Kind       ProjectKind
	Root       string
	Name       string
	CrateCount int    // Rust workspaces
	GitBranch  string
	GitAhead   int
	GitBehind  int
}

// DetectProject scans the working directory for project markers.
func DetectProject() *ProjectInfo {
	dir, err := os.Getwd()
	if err != nil {
		return nil
	}

	info := &ProjectInfo{Root: dir}

	// Check in order of priority
	if fileExists("Cargo.toml") {
		info.Kind = ProjectRust
		info.Name = extractCargoName()
		info.CrateCount = countCrates()
	} else if fileExists("package.json") {
		info.Kind = ProjectNode
		info.Name = extractPackageName()
	} else if fileExists("go.mod") {
		info.Kind = ProjectGo
		info.Name = extractGoModName()
	} else if fileExists("pyproject.toml") || fileExists("setup.py") || fileExists("requirements.txt") {
		info.Kind = ProjectPython
		info.Name = filepath.Base(dir)
	} else {
		info.Kind = ProjectUnknown
		info.Name = filepath.Base(dir)
	}

	// Git info
	info.GitBranch = detectGitBranch()
	info.GitAhead, info.GitBehind = detectGitAheadBehind()

	return info
}

// TestCommand returns the appropriate test command for the project type.
func (p *ProjectInfo) TestCommand() (string, []string) {
	switch p.Kind {
	case ProjectRust:
		return "cargo", []string{"test", "-j", "1"}
	case ProjectNode:
		return "npm", []string{"test"}
	case ProjectGo:
		return "go", []string{"test", "./..."}
	case ProjectPython:
		return "python", []string{"-m", "pytest"}
	}
	return "echo", []string{"No test command detected"}
}

// BuildCommand returns the build command for the project type.
func (p *ProjectInfo) BuildCommand() (string, []string) {
	switch p.Kind {
	case ProjectRust:
		return "cargo", []string{"build", "-j", "1"}
	case ProjectNode:
		return "npm", []string{"run", "build"}
	case ProjectGo:
		return "go", []string{"build", "./..."}
	case ProjectPython:
		return "python", []string{"-m", "build"}
	}
	return "echo", []string{"No build command detected"}
}

// LintCommand returns the lint command for the project type.
func (p *ProjectInfo) LintCommand() (string, []string) {
	switch p.Kind {
	case ProjectRust:
		return "cargo", []string{"clippy", "-j", "1"}
	case ProjectNode:
		return "npx", []string{"eslint", "."}
	case ProjectGo:
		return "golangci-lint", []string{"run"}
	case ProjectPython:
		return "ruff", []string{"check", "."}
	}
	return "echo", []string{"No lint command detected"}
}

// FmtCommand returns the format command for the project type.
func (p *ProjectInfo) FmtCommand() (string, []string) {
	switch p.Kind {
	case ProjectRust:
		return "cargo", []string{"fmt"}
	case ProjectNode:
		return "npx", []string{"prettier", "--write", "."}
	case ProjectGo:
		return "gofmt", []string{"-w", "."}
	case ProjectPython:
		return "ruff", []string{"format", "."}
	}
	return "echo", []string{"No format command detected"}
}

func fileExists(name string) bool {
	_, err := os.Stat(name)
	return err == nil
}

func extractCargoName() string {
	data, err := os.ReadFile("Cargo.toml")
	if err != nil {
		return "rust-project"
	}
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "name") && strings.Contains(line, "=") {
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				return strings.Trim(strings.TrimSpace(parts[1]), "\"")
			}
		}
	}
	return "rust-project"
}

func extractPackageName() string {
	data, err := os.ReadFile("package.json")
	if err != nil {
		return "node-project"
	}
	// Simple extraction — avoid full JSON parse
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "\"name\"") {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) == 2 {
				return strings.Trim(strings.TrimSpace(parts[1]), "\",")
			}
		}
	}
	return "node-project"
}

func extractGoModName() string {
	data, err := os.ReadFile("go.mod")
	if err != nil {
		return "go-project"
	}
	for _, line := range strings.Split(string(data), "\n") {
		if strings.HasPrefix(line, "module ") {
			return strings.TrimPrefix(line, "module ")
		}
	}
	return "go-project"
}

func countCrates() int {
	entries, err := os.ReadDir("crates")
	if err != nil {
		return 0
	}
	count := 0
	for _, e := range entries {
		if e.IsDir() {
			count++
		}
	}
	return count
}

func detectGitAheadBehind() (int, int) {
	out, err := exec.Command("git", "rev-list", "--left-right", "--count", "HEAD...@{upstream}").Output()
	if err != nil {
		return 0, 0
	}
	parts := strings.Fields(strings.TrimSpace(string(out)))
	if len(parts) != 2 {
		return 0, 0
	}
	ahead, behind := 0, 0
	_, _ = fmt.Sscanf(parts[0], "%d", &ahead)
	_, _ = fmt.Sscanf(parts[1], "%d", &behind)
	return ahead, behind
}
